src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+        let mut rng = CsRng::from_entropy();
+        let findex_master_key = KeyingMaterial::<MASTER_KEY_LENGTH>::new(&mut rng);
+
+        Ok(Token {
+            index_id,
+            findex_master_key,
+            fetch_entries_key: Some(fetch_entries_key),
+            fetch_chains_key: Some(fetch_chains_key),
+            upsert_entries_key: Some(upsert_entries_key),
+            insert_chains_key: Some(insert_chains_key),
+        })
+    }
+
+    pub fn reduce_permissions(&mut self, search: bool, index: bool) -> Result<(), FindexErr> {
+        self.fetch_entries_key =
+            reduce_option("fetch entries", self.fetch_entries_key, search || index)?;
+        self.fetch_chains_key = reduce_option("fetch chains", self.fetch_chains_key, search)?;
+        self.upsert_entries_key = reduce_option("upsert entries", self.upsert_entries_key, index)?;
+        self.insert_chains_key = reduce_option("insert chains", self.insert_chains_key, index)?;
+
+        Ok(())
+    }
+}
+
+fn reduce_option(
+    debug_info: &str,
+    permission: Option<[u8; 16]>,
+    keep: bool,
+) -> Result<Option<[u8; 16]>, FindexErr> {
+    if let Some(permission) = permission {
+        if keep {
+            Ok(Some(permission))
+        } else {
+            Err(FindexErr::Other(format!(
+                "The token provided doesn't have the permission to {debug_info}"
+            )))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+enum Callback {
+    FetchEntries,
+    FetchChains,
+    UpsertEntries,
+    InsertChains,
+}
+
+impl FindexCloud {
+    pub fn new(token: String, base_url: Option<String>) -> Result<Self, FindexErr> {
+        Ok(FindexCloud {
+            token: Token::from_str(&token)?,
+            base_url,
+        })
+    }
+
+    async fn post(&self, callback: Callback, bytes: &[u8]) -> Result<Vec<u8>, FindexErr> {
+        let endpoint = match callback {
+            Callback::FetchEntries => "fetch_entries",
+            Callback::FetchChains => "fetch_chains",
+            Callback::UpsertEntries => "upsert_entries",
+            Callback::InsertChains => "insert_chains",
+        };
+
+        let key = match callback {
+            Callback::FetchEntries => self.token.fetch_entries_key,
+            Callback::FetchChains => self.token.fetch_chains_key,
+            Callback::UpsertEntries => self.token.upsert_entries_key,
+            Callback::InsertChains => self.token.insert_chains_key,
+        }
+        .ok_or(FindexErr::Other(format!(
+            "your key doesn't have the permission to call {endpoint}"
+        )))?;
+
+        let mut hasher = Kmac::v128(&key, &[]);
+        let mut output = [0u8; 32];
+        hasher.update(bytes);
+        hasher.finalize(&mut output);
+
+        let signature = hex::encode(output);
+
+        let url = format!(
+            "{}/indexes/{}/{endpoint}",
+            self.base_url
+                .as_deref()
+                .unwrap_or("https://findex.cosmian.com"),
```

Comment:
I agree there may be something to do here. We may not want to hard code this URL.

Reply:
It's not really hard-coded since it's only the default value?

Reply:
True, but why is this value the default one ? Are we sure we will not change it?

I actually think it might be safer to remove the option at once.

Reply:
We're building a SAAS offering, it'll have a default domain (maybe not this one but it could be). Custom domain is only supported for development purposes (changing it to https://findex.staging.cosmian.com) or for on premise deployment in the future. 

Having the default value in Rust avoid copy pasting it in all clients, if we want to change it (for exemple for a Findex v2) we have a single source of truth (but it should be a constant)

Reply:
Okay, up to you :)

Reply:
ok good for me